frame/two-phase-election-provider/src/lib.rs

```diff
@@ -0,0 +1,1627 @@
+			let size = Self::build_solution_size().unwrap_or_default();
+
+			// ensure solution claims is better.
+			let mut signed_submissions = Self::signed_submissions();
+			let maybe_index =
+				Self::insert_submission(&who, &mut signed_submissions, solution, size);
+			ensure!(maybe_index.is_some(), Error::<T>::QueueFull);
+			let index = maybe_index.expect("Option checked to be `Some`; qed.");
+
+			// collect deposit. Thereafter, the function cannot fail.
+			// Defensive -- index is valid.
+			let deposit = signed_submissions
+				.get(index)
+				.map(|s| s.deposit)
+				.unwrap_or_default();
+			T::Currency::reserve(&who, deposit).map_err(|_| Error::<T>::CannotPayDeposit)?;
+
+			// store the new signed submission.
+			debug_assert!(signed_submissions.len() as u32 <= T::MaxSignedSubmissions::get());
+			log!(
+				info,
+				"queued signed solution with (claimed) score {:?}",
+				signed_submissions
+					.get(index)
+					.map(|s| s.solution.score)
+					.unwrap_or_default()
+			);
+
+			<SignedSubmissions<T>>::put(signed_submissions);
+			Self::deposit_event(Event::SolutionStored(ElectionCompute::Signed));
+			Ok(None.into())
+		}
+
+		/// Submit a solution for the unsigned phase.
+		///
+		/// The dispatch origin fo this call must be __none__.
+		///
+		/// This submission is checked on the fly, thus it is likely yo be more limited and smaller.
+		/// Moreover, this unsigned solution is only validated when submitted to the pool from the
+		/// local process. Effectively, this means that only active validators can submit this
+		/// transaction when authoring a block.
+		///
+		/// To prevent any incorrect solution (and thus wasted time/weight), this transaction will
+		/// panic if the solution submitted by the validator is invalid, effectively putting their
+		/// authoring reward at risk.
+		///
+		/// No deposit or reward is associated with this.
+		#[pallet::weight(T::WeightInfo::submit_unsigned(
+			witness.voters,
+			witness.targets,
+			solution.compact.voter_count() as u32,
+			solution.compact.unique_targets().len() as u32
+		))]
+		pub fn submit_unsigned(
+			origin: OriginFor<T>,
+			solution: RawSolution<CompactOf<T>>,
+			witness: SolutionSize,
+		) -> DispatchResultWithPostInfo {
+			ensure_none(origin)?;
+			let error_message = "Invalid unsigned submission must produce invalid block and \
+				deprive validator from their authoring reward.";
+
+			// check phase and score.
+			// NOTE: since we do this in pre-dispatch, we can just ignore it here.
+			Self::unsigned_pre_dispatch_checks(&solution).expect(error_message);
+
+			// ensure witness was correct.
+			let RoundSnapshotMetadata {
+				voters_len,
+				targets_len,
+			} = Self::snapshot_metadata().expect(error_message);
+
+			// NOTE: we are asserting, not `ensure`ing -- we want to panic here.
+			assert!(voters_len as u32 == witness.voters, error_message);
+			assert!(targets_len as u32 == witness.targets, error_message);
```

Comment:
What happens if someone modifies their runtime or makes a block or something which does not have this assert statement.

Then this block is shared to real validators, and when they run this extrinsic themself they will panic?

maybe i have some knowledge of the flow or what is possible wrong.

Reply:
I think the rule of never panic in the runtime is deeper than what you are doing here

Reply:
The runtime is stored on chain and thus cannot be modified. An author is free to modify their runtime and make a transaction that _should have_ failed on these assertions, but all other validators will panic, which means that their block will not be imported, thus no reward for them. 

> Then this block is shared to real validators, and when they run this extrinsic themself they will panic?

So yes this is the intention more or less. 

> 

I think the rule of never panic in the runtime is deeper than what you are doing here

Happy to hear more, to me, it makes sense to panic here.

Reply:
If an extrinsic panics, is it correctly flushed by the transaction queue? I would imagine so...

I dunno, this feels like an aggressive measure to me to panic here, but maybe it is okay

Reply:
> If an extrinsic panics, is it correctly flushed by the transaction queue? I would imagine so..

The answer is yes but either way this transaction never hits the transaction queue: It is only allowed when it has `source::Local` + it has `.propagate(false)`, meaning that it is only valid only if it comes from the local node, which means you must be an author. Thus, it is basically like an inherent, we just prefer to make it an unsigned tx. 

InitiaInitial discussion: 
https://github.com/paritytech/substrate/pull/4517#issuecomment-602526376